src/goto-harness/function_call_harness_generator.cpp

```diff
@@ -53,82 +99,147 @@ void function_call_harness_generatort::generate(
+    auto argument = allocate_objects.allocate_automatic_local_object(
+      parameter.type(), parameter.get_base_name());
+    arguments.push_back(argument);
   }
-
-  if(harness_function_found != nullptr)
+  allocate_objects.declare_created_symbols(function_body);
+  for(auto const &argument : arguments)
   {
-    throw invalid_command_line_argument_exceptiont{
-      "harness function already in the symbol table " +
-        id2string(harness_function_name),
-      "--" GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT};
+    generate_initialisation_code_for(function_body, argument);
   }
+  code_function_callt function_call{function_to_call.symbol_expr(),
+                                    std::move(arguments)};
+  function_call.add_source_location() = function_to_call.location;
 
-  auto allocate_objects = allocate_objectst{function_found->mode,
-                                            function_found->location,
-                                            "__goto_harness",
-                                            symbol_table};
+  function_body.add(std::move(function_call));
+}
+
+void function_call_harness_generatort::implt::generate(
+  goto_modelt &goto_model,
+  const irep_idt &harness_function_name)
+{
+  symbol_table = &goto_model.symbol_table;
+  goto_functions = &goto_model.goto_functions;
+  const auto &function_to_call = lookup_function_to_call();
+  recursive_initialization_config.mode = function_to_call.mode;
+  recursive_initialization = util_make_unique<recursive_initializationt>(
+    recursive_initialization_config, goto_model, *message_handler);
+  this->harness_function_name = harness_function_name;
+  ensure_harness_does_not_already_exist();
 
   // create body for the function
   code_blockt function_body{};
 
-  const auto &function_type = to_code_type(function_found->type);
-  const auto &parameters = function_type.parameters();
+  generate_nondet_globals(function_body);
+  setup_parameters_and_call_entry_function(function_body);
+  add_harness_function_to_goto_model(std::move(function_body));
+}
 
-  code_function_callt::operandst arguments{};
-  arguments.reserve(parameters.size());
+void function_call_harness_generatort::implt::generate_nondet_globals(
+  code_blockt &function_body)
+{
+  if(nondet_globals)
+  {
+    for(const auto &symbol_table_entry : *symbol_table)
+    {
+      const auto &symbol = symbol_table_entry.second;
+      if(
+        symbol.is_static_lifetime && symbol.is_lvalue &&
+        !has_prefix(id2string(symbol.name), CPROVER_PREFIX))
+      {
+        generate_initialisation_code_for(function_body, symbol.symbol_expr());
+      }
+    }
+  }
+}
 
-  for(const auto &parameter : parameters)
+void function_call_harness_generatort::implt::generate_initialisation_code_for(
+  code_blockt &block,
+  const exprt &lhs)
+{
+  recursive_initialization->initialize(lhs, 0, block);
+}
+
+void function_call_harness_generatort::validate_options()
```

Comment:
Is this one called anywhere?

Reply:
This is called by the harness generator factory